tests/importer/importer.cpp

```diff
@@ -0,0 +1,587 @@
+    //     std::cout << importer->externalDependency(e).first << " " << importer->externalDependency(e).second << std::endl;
+    // }
+
+    // Now create a funky model that imports from all over the place and see what happens.
+    std::string funkyString = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+                              "<model xmlns=\"http://www.cellml.org/cellml/2.0#\" name=\"funky\">\n"
+                              // Status quo functionality, key is local URL + base file path as the local URL doesn't exist as a key.
+                              "  <import xmlns:xlink=\"http://www.w3.org/1999/xlink\" xlink:href=\"units_imported.cellml\">\n"
+                              "    <units units_ref=\"units1_imported\" name=\"units1FromLibrary\"/>\n"
+                              "  </import>\n"
+                              // Use the local URL for a hidden child source component.
+                              "  <import xmlns:xlink=\"http://www.w3.org/1999/xlink\" xlink:href=\"components_source.cellml\">\n"
+                              "    <component component_ref=\"component1\" name=\"component1FromLibrary\"/>\n"
+                              "  </import>\n"
+                              // Use the absolute URL for a hidden child source component.
+                              "  <import xmlns:xlink=\"http://www.w3.org/1999/xlink\" xlink:href=\""
+                              + resourcePath("resolveimports/components_source.cellml")
+                              + "\">\n"
+                                "    <component component_ref=\"component2\" name=\"component2FromLibrary\"/>\n"
+                                "  </import>\n"
+                                // Remote website URL used for key.
+                                "  <import xmlns:xlink=\"http://www.w3.org/1999/xlink\" xlink:href=\"https://www.example.com/myComponents.cellml\">\n"
+                                "    <component component_ref=\"component3\" name=\"component3FromLibrary\"/>\n"
+                                "  </import>\n"
+                                // Random string used for key. Most certainly not a valid URL.
+                                "  <import xmlns:xlink=\"http://www.w3.org/1999/xlink\" xlink:href=\"the concrete units file\">\n"
+                                "    <units units_ref=\"units4\" name=\"units4FromLibrary\"/>\n"
+                                "  </import>\n"
+                                // Numerical key.
+                                "  <import xmlns:xlink=\"http://www.w3.org/1999/xlink\" xlink:href=\"1234567890\">\n"
+                                "    <units units_ref=\"units5\" name=\"units5FromLibrary\"/>\n"
+                                "  </import>\n"
+                                // Key with special characters.
+                                "  <import xmlns:xlink=\"http://www.w3.org/1999/xlink\" xlink:href=\"#hello!\">\n"
```

Comment:
important to remember that `xlink:href` is not a key, it is a well defined thing in XLink...i.e., a URL. Pretty sure `#hello!` is invalid, at least the `!` would need to be URL-encoded. And these URLs must be resolved in order to resolve them.

Reply:
According to our current setup, these are all valid ... we only check for presence at the moment.  Might need to change that then?  And the URLs don't need to be resolved ... that's the whole point.  The importer treats them as keys, not urls.  They're only urls when the key doesn't work.  If we don't use that approach, then users who manually add models to the importer (which is what I'd thought its main functionality was ...), they will have to specify the absolute URL.  This makes the code workstation-dependent, and really restricted? I don't understand the motivation for that?

Reply:
Imports cannot be resolved without resolving the href URL. This is essential to the way that imports are defined and how they work, and the definition of the `xlink:href` property. Relative URLs are recommended for serialisation to avoid the workstation dependence for local files, but they still must be resolved into absolute URLs by processing software before they can be interpreted. They are definitely not just keys. Just think of a CellML document stored in `model.xml`. I then import from `components/model.xml` and that in turn imports from its own `components/model.xml` - if these relative URLs are not resolved to their absolute form for the current context then you have no idea which `model.xml` is which.

Reply:
When you resolve a model's imports, those imports are added to the library as external dependencies, under the absolute URL at which they were found.  This means that there's no conflict in your example, both the `someLocalPath/model.xml` and `someLocalPath/components/model.xml` are stored separately, as you'd expect.  All my implementation does is extend this to allow manually created keys (where they are keys to manually-supplied models) as well.  Maybe the keys are better named "aliases"?